File: boincsite/status/rpc/GlobalPreferences.py
# Copyright (c) David Wilson 2016
#
# Licensed under the GPL version 3

import json

attrs = [
   'source_project',
   'mod_time',
   'battery_charge_min_pct',
   'battery_max_temperature',
   'run_on_batteries',
   'run_if_user_active',
   'run_gpu_if_user_active',
   'suspend_if_no_recent_input',
   'suspend_cpu_usage',
   'start_hour',
   'end_hour',
   'net_start_hour',
   'net_end_hour',
   'leave_apps_in_memory',
   'confirm_before_connecting',
   'hangup_if_dialed',
   'dont_verify_images',
   'work_buf_min_days',
   'work_buf_additional_days',
   'max_ncpus_pct',
   'cpu_scheduling_period_minutes',
   'disk_interval',
   'disk_max_used_gb',
   'disk_max_used_pct',
   'disk_min_free_gb',
   'vm_max_used_pct',
   'ram_max_used_busy_pct',
   'ram_max_used_idle_pct',
   'idle_time_to_run',
   'max_bytes_sec_up',
   'max_bytes_sec_down',
   'cpu_usage_limit',
   'daily_xfer_limit_mb',
   'daily_xfer_period_days',
   'override_file_present',
   'network_wifi_only',
   'max_cpus'
]

class GlobalPreferences(object):

    def __init__(self, global_preferences):
        for a in attrs:
            setattr(self, a, getattr(global_preferences, a))

    def __str__(self):
        return '''Source Project: {source_project}
Mod time: {mod_time}
Battery charge minimum%: {battery_charge_min_pct}
Battery Max Temprature: {battery_max_temperature}
Run on batteries: {run_on_batteries}
Run if user active: {run_if_user_active}
Run GPU if user active: {run_gpu_if_user_active}
Suspend if no recent input: {suspend_if_no_recent_input}
Suspend CPU usage: {suspend_cpu_usage}
Start hour: {start_hour}
End hour: {end_hour}
Net start hour: {net_start_hour}
Net end hour: {net_end_hour}
Leave apps in memory: {leave_apps_in_memory}
Confirm before connecting: {confirm_before_connecting}
Hangup if dialed: {hangup_if_dialed}
Don't verify images: {dont_verify_images}
Work buffer minimum days: {work_buf_min_days}
Work buffer additional days: {work_buf_additional_days}
max_ncpus_pct: {max_ncpus_pct}
CPU scheduling period minutes: {cpu_scheduling_period_minutes}
Disk interval: {disk_interval}
Disk max used gigabytes: {disk_max_used_gb}
Disk max used%: {disk_max_used_pct}
Disk min free gb: {disk_min_free_gb},
vm_max_used_pct: {vm_max_used_pct},
Ram max used busy pct: {ram_max_used_busy_pct}
Ram max used idle pct: {ram_max_used_idle_pct}
Idle time to run: {idle_time_to_run}
Max bytes/second up: {max_bytes_sec_up}
Max bytes/second down: {max_bytes_sec_down}
CPU usage limit: {cpu_usage_limit}
Daily transfer limit MB: {daily_xfer_limit_mb}
Daily transfer period (days): {daily_xfer_period_days}
Override file present: {override_file_present}
Network wifi only: {network_wifi_only}
Max CPUs: {max_cpus}
'''.format(source_project=self.source_project,
           mod_time=self.mod_time,
           battery_charge_min_pct=self.battery_charge_min_pct,
           battery_max_temperature=self.battery_max_temperature,
           run_on_batteries=self.run_on_batteries,
           run_if_user_active=self.run_if_user_active,
           run_gpu_if_user_active=self.run_gpu_if_user_active,
           suspend_if_no_recent_input=self.suspend_if_no_recent_input,
           suspend_cpu_usage=self.suspend_cpu_usage,
           start_hour=self.start_hour,
           end_hour=self.end_hour,
           net_start_hour=self.net_start_hour,
           net_end_hour=self.net_end_hour,
           leave_apps_in_memory=self.leave_apps_in_memory,
           confirm_before_connecting=self.confirm_before_connecting,
           hangup_if_dialed=self.hangup_if_dialed,
           dont_verify_images=self.dont_verify_images,
           work_buf_min_days=self.work_buf_min_days,
           work_buf_additional_days=self.work_buf_additional_days,
           max_ncpus_pct=self.max_ncpus_pct,
           cpu_scheduling_period_minutes=self.cpu_scheduling_period_minutes,
           disk_interval=self.disk_interval,
           disk_max_used_gb=self.disk_max_used_gb,
           disk_max_used_pct=self.disk_max_used_pct,
           disk_min_free_gb=self.disk_min_free_gb,
           vm_max_used_pct=self.vm_max_used_pct,
           ram_max_used_busy_pct=self.ram_max_used_busy_pct,
           ram_max_used_idle_pct=self.ram_max_used_idle_pct,
           idle_time_to_run=self.idle_time_to_run,
           max_bytes_sec_up=self.max_bytes_sec_up,
           max_bytes_sec_down=self.max_bytes_sec_down,
           cpu_usage_limit=self.cpu_usage_limit,
           daily_xfer_limit_mb=self.daily_xfer_limit_mb,
           daily_xfer_period_days=self.daily_xfer_period_days,
           override_file_present=self.override_file_present,
           network_wifi_only=self.network_wifi_only,
           max_cpus=self.max_cpus)


class JSONEncoder(json.JSONEncoder):
    def default(self, o):
        d = {}

        for a in attrs:
            d[a] = getattr(o, a)

        return d

File: boincsite/zoetrope-client/src/components/globalpreferences.js
import React, { useEffect, useState } from 'react';

import { getGlobalPreferences } from '../api/Api';
import { FourColumnRow } from './primitives/row';

const GlobalPreferences = () => {

  const [globalPrefs, setGlobalPrefs] = useState({});

  useEffect(() => {
    getGlobalPreferences(x => setGlobalPrefs(x));
  }, []);

  const MakeRow = ({label1, key1, label2, key2}) => {
    const col1 = { col1Label: label1, col1Value: globalPrefs[key1] };
    const col2 = { col2Label: label2, col2Value: globalPrefs[key2] };

    return FourColumnRow({col1, col2});
  };

  return (
    <>
    <h2>Global Preferences</h2>
    <h3>CPU/GPU</h3>
    <MakeRow label1="Percentage of the CPUs to use" key1={'cpu_usage_limit'} 
      label2="Maximum CPU Usage" key2={'suspend_cpu_usage'} />
    <MakeRow label1="Maximum number of CPUs to use" key1={'max_cpus'} 
      label2="Use GPU if computer in use" key2={'run_gpu_if_user_active'} />
    <MakeRow label1="CPU scheduling period" key1={'cpu_scheduling_period_minutes'} 
      label2="" key2={''} />

    <h3>Memory</h3>
    <MakeRow label1="Maximum RAM to use when idle" key1={'ram_max_used_idle_pct'} 
      label2="Maximum % of RAM to use when in use" key2={'ram_max_used_busy_pct'} />
    <MakeRow label1="Leave apps in memory?" key1={'leave_apps_in_memory'} 
      label2="" key2={''} />

    <h3>When to Run</h3>
    <MakeRow label1="Start hour" key1={'start_hour'} 
      label2="End hour" key2={'end_hour'} />
    <MakeRow label1="Run if computer in use" key1={'run_if_user_active'} 
      label2="Run when idle for" key2={'idle_time_to_run'} />

    <h3>Network</h3>
    <MakeRow label1="Hangup if dialed" key1={'hangup_if_dialed'} 
      label2="Confirm before making an internet connection" key2={'idle_time_to_run'} />
    <MakeRow label1="Maximum download speed" key1={'max_byes_sec_down'} 
      label2="Maximum upload speed" key2={'max_byes_sec_up'} />
    <MakeRow label1="Starting hour for network traffic" key1={'net_start_hour'} 
      label2="Ending hour for network traffic" key2={'net_end_hour'} />
    <MakeRow label1="WiFi only?" key1={'network_wifi_only'} 
      label2="" key2={''} />

    <h3>Battery</h3>
    <MakeRow label1="Maximum Battery temperature" key1={'battery_max_temperature'} 
      label2="Minimum battery charge" key2={'battery_charge_min_pct'} />
    <MakeRow label1="Run on batteries?" key1={'run_on_batteries'} 
      label2="" key2={''} />

    <h3>Disk</h3>
    <MakeRow label1="Maximum disk space to use" key1={'disk_max_used_gb'} 
      label2="Maximum percentage diks space to use" key2={'disk_max_used_pct'} />
    <MakeRow label1="Make task checkpoint to disk every (seconds)" key1={'disk_interval'} 
      label2="Minimum disk space to leave free" key2={'disk_min_free_gb'} />

    <h3>Misc. Settings</h3>
    <MakeRow label1="daily_xfer_period_days" key1={'daily_xfer_period_days'} 
      label2="Don't verify images" key2={'dont_verify_images'} />
    <MakeRow label1="mod_time" key1={'mod_time'} 
      label2="Override file present" key2={'override_file_present'} />
    <MakeRow label1="Source Project" key1={'source_project'} 
      label2="suspend_if_no_recent_input" key2={'suspend_if_no_recent_input'} />
    <MakeRow label1="Amount of extra work to buffer" key1={'work_buf_additional_days'} 
      label2="Minimum work to buffer" key2={'work_buf_min_days'} />
    <MakeRow label1="vm_max_used_pct" key1={'vm_max_used_pct'} 
      label2="" key2={''} />


    </>
  );
};

export default GlobalPreferences;

